File: src/tools/elevation/contours.ts
import {
  buildSqDist,
  convertMapboxElevationData,
  getElevationGrid,
  polygonRingArea,
  polylineInPolyline,
  simplifyLine,
} from '../../index.js';

import type {
  ElevationConverter,
  MValue,
  Properties,
  VectorFeatureCollection,
  VectorLineString,
  VectorMultiLineString,
  VectorMultiPolygon,
  VectorMultiPolygonFeature,
  VectorMultiPolygonGeometry,
  VectorPoint,
} from '../../index.js';

export interface ElevationProperties extends Properties {
  elev: number;
  elevFt: number;
}

export type ContourFeature = VectorMultiPolygonFeature<
  Record<string, unknown>,
  MValue,
  ElevationProperties
>;

export type ContourFeatureCollection = VectorFeatureCollection<
  Record<string, unknown>,
  MValue,
  ElevationProperties,
  VectorMultiPolygonGeometry
>;

interface Fragment {
  start: number;
  end: number;
  ring: VectorLineString;
}

// prettier-ignore
const MS_LUT = [
  [], // 0
  [[{ x: 1.0, y: 1.5 }, { x: 0.5, y: 1.0 }]], // 1
  [[{ x: 1.5, y: 1.0 }, { x: 1.0, y: 1.5 }]], // 2
  [[{ x: 1.5, y: 1.0 }, { x: 0.5, y: 1.0 }]], // 3
  [[{ x: 1.0, y: 0.5 }, { x: 1.5, y: 1.0 }]], // 4
  [[{ x: 1.0, y: 1.5 }, { x: 0.5, y: 1.0 }],[{ x: 1.0, y: 0.5 }, { x: 1.5, y: 1.0 }]], // 5
  [[{ x: 1.0, y: 0.5 }, { x: 1.0, y: 1.5 }]], // 6
  [[{ x: 1.0, y: 0.5 }, { x: 0.5, y: 1.0 }]], // 7
  [[{ x: 0.5, y: 1.0 }, { x: 1.0, y: 0.5 }]], // 8
  [[{ x: 1.0, y: 1.5 }, { x: 1.0, y: 0.5 }]], // 9
  [[{ x: 0.5, y: 1.0 }, { x: 1.0, y: 0.5 }],[{ x: 1.5, y: 1.0 }, { x: 1.0, y: 1.5 }]], // 10
  [[{ x: 1.5, y: 1.0 }, { x: 1.0, y: 0.5 }]], // 11
  [[{ x: 0.5, y: 1.0 }, { x: 1.5, y: 1.0 }]], // 12
  [[{ x: 1.0, y: 1.5 }, { x: 1.5, y: 1.0 }]], // 13
  [[{ x: 0.5, y: 1.0 }, { x: 1.0, y: 1.5 }]], // 14
  [], // 15
];

/**
 * Generate isoline thresholds given a min, max, and step.
 * @param min - minimum value
 * @param max - maximum value
 * @param step - step size
 * @returns a collection of thresholds within the range provided all inclusive and sorted
 */
export function isolineThresholds(min: number, max: number, step: number): number[] {
  const thresholds: number[] = [];
  let current = Math.ceil(min / step) * step;

  while (current <= max) {
    thresholds.push(current === 0 ? 0 : current); // avoid -0
    current += step;
  }

  return thresholds;
}

/**
 * # Create Isolines
 *
 * ## Description
 * Creates isolines from an image.
 *
 * NOTE: Defaults to the Mapbox elevation data converter `convertMapboxElevationData`. However,
 * to use the Terrarium elevation data converter, use `convertTerrariumElevationData`.
 *
 * NOTE: Using a buffer with a small padding works, but it only extends the line ends of the isolines.
 * A better method would be to pull in grid data with large padding (second example).
 * If you have a large padding like 16px then you don't need a buffer, so you should set it to 0.
 *
 * ## Examples
 *
 * ### Create Isolines
 * ```ts
 * import 'gis-tools-ts/polyfills/local'; // You may need this to handle image data if you are using the file system.
 * import { buildContours } from 'gis-tools-ts';
 *
 * // Pull in the image data. We are using a local image file that is 514x514 with prebuilt padding
 * const elevationImage = await Bun.file(
 *   './tests/tools/isobands/fixtures/13_1556_3084.webp',
 * ).arrayBuffer();
 *
 * // Create the isolines.
 * const isolines = await buildContours(elevationImage);
 * ```
 *
 * ### High quality with large padding
 * ```ts
 * import { RasterTilesFileReader } from 'gis-tools-ts/file'; // because we use file.js local polyfills already added
 * import { buildContours } from 'gis-tools-ts';
 *
 * // Setup a tile reader
 * const reader = new RasterTilesFileReader(
 *   `${__dirname}/../../readers/tile/fixtures/wm/terrain-v2`,
 * );
 * const metadata = await reader.getMetadata();
 * const isTMS = metadata.scheme === 'tms';
 * // read in the elevation data with a big bigger padding
 * // we are pulling in Canada/Greenland area at zoom 3. Padding of 16 so a resulting 544x544 image.
 * const padding = 16;
 * const tile = await reader.getTileWithPaddingWM(3, 3, 1, padding, 512, 512);
 * if (tile === undefined) throw new Error('Tile not found');
 * // build the isolines
 * const isolines = await buildContours(tile!.image, undefined, 1_000, isTMS, padding, 0);
 * ```
 *
 * @param imageData - the raw RGB(A) image data
 * @param elevationConverter - the conversion function to convert the pixels to elevation
 * @param step - the step size for the heightmap. Defaults to 100 meters for the Mapbox elevation data.
 * @param tmsStyle - if true, the y position will be inverted
 * @param padding - The number of pixels that extend around the main data
 * @param tolerance - The Douglas-Peucker tolerance. Defaults to `1 / 2_096`
 * @returns The isolines stored in a FeatureCollection
 */
export async function buildContours(
  imageData:
    | ArrayBufferLike
    | Uint8Array<ArrayBufferLike>
    | Uint8ClampedArray<ArrayBufferLike>
    | Buffer<ArrayBufferLike>
    | ImageData,
  elevationConverter: ElevationConverter = convertMapboxElevationData,
  step = 100,
  tmsStyle = false,
  padding = 1,
  tolerance = 1 / 2_096,
): Promise<ContourFeatureCollection> {
  // create the elevation grid
  const elevationGrid = await getElevationGrid(imageData, elevationConverter, tmsStyle);
  const { width, height, min, max, elevations } = elevationGrid;
  // setup the thresholds
  const thresholds = isolineThresholds(min, max, step);

  const features: ContourFeature[] = [];
  for (const threshold of thresholds) {
    const coordinates = buildIsoBands(elevations, threshold, width, height, padding, tolerance);
    features.push({
      type: 'VectorFeature',
      properties: { elev: threshold, elevFt: threshold * 3.28084 },
      geometry: { type: 'MultiPolygon', is3D: false, coordinates },
    });
  }

  return { type: 'FeatureCollection', features };
}

// Accumulate, smooth contour rings, assign holes to exterior rings.
// Based on https://github.com/mbostock/shapefile/blob/v0.6.2/shp/polygon.js
// @param tolerance - The Douglas-Peucker tolerance. A good default is `1 / 2_096`
export function buildIsoBands(
  values: number[],
  threshold: number,
  width: number,
  height: number,
  padding: number,
  tolerance = 0,
): VectorMultiPolygon {
  const polygons: VectorMultiPolygon = [];
  const holes: VectorLineString[] = [];

  const isorings = buildIsorings(values, threshold, width, height, padding, tolerance);
  // Store rings in the correct groups. Skip rings with zero area.
  for (const ring of isorings) {
    const area = polygonRingArea(ring, 1);
    if (area === 0) continue;
    else if (area > 0) polygons.push([ring]);
    else holes.push(ring);
  }
  // sort the holes into their correct polygons
  for (const hole of holes) {
    for (const polygon of polygons) {
      if (polylineInPolyline(hole, polygon[0])) {
        polygon.push(hole);
        break;
      }
    }
  }

  return polygons;
}

// Marching squares with isolines stitched into rings.
// Based on https://github.com/topojson/topojson-client/blob/v3.0.0/src/stitch.js
// @param tolerance - The Douglas-Peucker tolerance. A good default is `1 / 2_096`
export function buildIsorings(
  values: number[],
  threshold: number,
  width: number,
  height: number,
  padding: number,
  tolerance = 0,
): VectorMultiLineString {
  const res: VectorMultiLineString = [];
  const fragByStart = new Map<number, Fragment>();
  const fragByEnd = new Map<number, Fragment>();
  let x: number;
  let y: number;
  let t0: number;
  let t1: number;
  let t2: number;
  let t3: number;
  // Convert a point to an index.
  const index = (point: VectorPoint, width: number): number => {
    return point.x * 2 + point.y * (width + 1) * 4;
  };
  // Convert to a number (0 or 1) whether the value is above the threshold.
  const above = (x: number | undefined, value: number): number => {
    return x === undefined ? 0 : Number(x >= value);
  };

  // Special case for the first row (y = -1, t2 = t3 = 0).
  x = y = -1;
  t1 = above(values[0], threshold);
  MS_LUT[t1 << 1].forEach(stitch);
  while (++x < width - 1) {
    t0 = t1;
    t1 = above(values[x + 1], threshold);
    MS_LUT[t0 | (t1 << 1)].forEach(stitch);
  }
  MS_LUT[t1 << 0].forEach(stitch);

  // General case for the intermediate rows.
  while (++y < height - 1) {
    x = -1;
    t1 = above(values[y * width + width], threshold);
    t2 = above(values[y * width], threshold);
    MS_LUT[(t1 << 1) | (t2 << 2)].forEach(stitch);
    while (++x < width - 1) {
      t0 = t1;
      t1 = above(values[y * width + width + x + 1], threshold);
      t3 = t2;
      t2 = above(values[y * width + x + 1], threshold);
      MS_LUT[t0 | (t1 << 1) | (t2 << 2) | (t3 << 3)].forEach(stitch);
    }
    MS_LUT[t1 | (t2 << 3)].forEach(stitch);
  }

  // Special case for the last row (y = height - 1, t0 = t1 = 0).
  x = -1;
  t2 = Number(values[y * width] >= threshold);
  MS_LUT[t2 << 2].forEach(stitch);
  while (++x < width - 1) {
    t3 = t2;
    t2 = above(values[y * width + x + 1], threshold);
    MS_LUT[(t2 << 2) | (t3 << 3)].forEach(stitch);
  }
  MS_LUT[t2 << 3].forEach(stitch);

  function stitch(line: VectorLineString): void {
    const start = { x: line[0].x + x, y: line[0].y + y };
    const end = { x: line[1].x + x, y: line[1].y + y };
    const startIndex = index(start, width);
    const endIndex = index(end, width);

    let f = fragByEnd.get(startIndex);
    let g = fragByStart.get(endIndex);

    if (f !== undefined) {
      if (g !== undefined) {
        fragByEnd.delete(f.end);
        fragByStart.delete(g.start);
        if (f === g) {
          f.ring.push(end);
          res.push(f.ring);
        } else {
          const frag = {
            start: f.start,
            end: g.end,
            ring: f.ring.concat(g.ring),
          };
          fragByStart.set(f.start, frag);
          fragByEnd.set(g.end, frag);
        }
      } else {
        fragByEnd.delete(f.end);
        f.ring.push(end);
        f.end = endIndex;
        fragByEnd.set(endIndex, f);
      }
    } else if ((f = fragByStart.get(endIndex)) !== undefined) {
      if ((g = fragByEnd.get(startIndex)) !== undefined) {
        fragByStart.delete(f.start);
        fragByEnd.delete(g.end);
        if (f === g) {
          f.ring.push(end);
          res.push(f.ring);
        } else {
          const frag = {
            start: g.start,
            end: f.end,
            ring: g.ring.concat(f.ring),
          };
          fragByStart.set(g.start, frag);
          fragByEnd.set(f.end, frag);
        }
      } else {
        fragByStart.delete(f.start);
        f.ring.unshift(start);
        f.start = startIndex;
        fragByStart.set(startIndex, f);
      }
    } else {
      const frag = { start: startIndex, end: endIndex, ring: [start, end] };
      fragByStart.set(startIndex, frag);
      fragByEnd.set(endIndex, frag);
    }
  }

  for (let i = 0; i < res.length; i++) {
    const ring = res[i];
    // reverse the ring
    ring.reverse();
    // smooth the edges
    smooth(ring, values, threshold, width, height);
    // convert points to 0->1
    for (const point of ring) remap(point, width, height, padding);
    // lastly simplify if tolerance is not 0
    if (tolerance !== 0) {
      // Prep Douglas-Peucker simplification by setting t-values.
      buildSqDist(ring, 0, ring.length - 1, tolerance * tolerance);
      // Apply Douglas-Peucker
      res[i] = simplifyLine(ring, tolerance, false, false);
    }
  }

  return res;
}

function smooth(
  ring: VectorLineString,
  values: number[],
  threshold: number,
  width: number,
  height: number,
): void {
  ring.forEach((point): void => {
    const { x, y } = point;
    const xt = x | 0;
    const yt = y | 0;
    const v1 = valid(values[yt * width + xt]);
    if (x > 0 && x < width && xt === x) {
      point.x = smooth1(x, valid(values[yt * width + xt - 1]), v1, threshold);
    }
    if (y > 0 && y < height && yt === y) {
      point.y = smooth1(y, valid(values[(yt - 1) * width + xt]), v1, threshold);
    }
  });
}

function valid(v: number | undefined): number {
  return v === undefined || isNaN(v) ? -Infinity : v;
}

function smooth1(x: number, v0: number, v1: number, value: number): number {
  const a = value - v0;
  const b = v1 - v0;
  const d = isFinite(a) || isFinite(b) ? a / b : Math.sign(a) / Math.sign(b);
  return isNaN(d) ? x : x + d - 0.5;
}

// Convert the x-y values from width-height scale to 0->1.
// Input data was already offest to represent cell centers ([0.5, 0.5] start)
function remap(point: VectorPoint, width: number, height: number, padding: number): void {
  const activeWidth = width - 2 * padding;
  const activeHeight = height - 2 * padding;
  point.x = (point.x - padding) / activeWidth;
  point.y = (point.y - padding) / activeHeight;
}

File: rust/tools/elevation/contours.rs
use crate::{
    geometry::{Area, build_sq_dist, polyline_in_polyline, simplify_line},
    parsers::Buffer,
    tools::{ElevationConverter, IsoGrid, get_elevation_grid},
};
use alloc::{collections::BTreeMap, rc::Rc};
use core::cell::RefCell;
use libm::{ceil, trunc};
use s2json::{
    FeatureCollection, Features, GetXY, MValue, SetXY, VectorFeature, VectorGeometry,
    VectorLineString, VectorMultiLineString, VectorMultiPolygon, VectorPoint,
};
use serde::{Deserialize, Serialize};

/// Contour Properties
#[derive(Debug, Default, Clone, Copy, PartialEq, Serialize, Deserialize, MValue)]
pub struct ContourProperties {
    /// Elevation in meters
    pub elev: f64,
    /// Elevation in feet
    #[serde(rename = "elevFt")]
    pub elev_ft: f64,
}

#[derive(Debug, Clone)]
struct Fragment {
    pub start: usize,
    pub end: usize,
    pub ring: VectorLineString,
}
impl Fragment {
    pub fn new(start: usize, end: usize, ring: VectorLineString) -> Rc<RefCell<Fragment>> {
        Rc::new(RefCell::new(Fragment { start, end, ring }))
    }
}

// Each index 0-15 returns pairs of edges to connect.
// Example: Case 3 (binary 0011) connects edge 3 to edge 1.
static MS_LUT: &[&[[(f64, f64); 2]]] = &[
    &[],                                                   // 0
    &[[(1.0, 1.5), (0.5, 1.0)]],                           // 1
    &[[(1.5, 1.0), (1.0, 1.5)]],                           // 2
    &[[(1.5, 1.0), (0.5, 1.0)]],                           // 3
    &[[(1.0, 0.5), (1.5, 1.0)]],                           // 4
    &[[(1.0, 1.5), (0.5, 1.0)], [(1.0, 0.5), (1.5, 1.0)]], // 5
    &[[(1.0, 0.5), (1.0, 1.5)]],                           // 6
    &[[(1.0, 0.5), (0.5, 1.0)]],                           // 7
    &[[(0.5, 1.0), (1.0, 0.5)]],                           // 8
    &[[(1.0, 1.5), (1.0, 0.5)]],                           // 9
    &[[(0.5, 1.0), (1.0, 0.5)], [(1.5, 1.0), (1.0, 1.5)]], // 10
    &[[(1.5, 1.0), (1.0, 0.5)]],                           // 11
    &[[(0.5, 1.0), (1.5, 1.0)]],                           // 12
    &[[(1.0, 1.5), (1.5, 1.0)]],                           // 13
    &[[(0.5, 1.0), (1.0, 1.5)]],                           // 14
    &[],                                                   // 15
];

/// Get the isoline thresholds relative to a minimum, maximum, and step size
///
/// ## Parameters
/// - `min`: The minimum elevation
/// - `max`: The maximum elevation
/// - `step`: The step size
///
/// ## Returns
/// An array of thresholds within the range provided
pub fn isoline_thresholds(min: f64, max: f64, step: f64) -> Vec<f64> {
    let mut thresholds: Vec<f64> = vec![];
    let mut current = ceil(min / step) * step;

    while current <= max {
        thresholds.push(current);
        current += step;
    }

    thresholds
}

/// # Create Isolines or Isobands
///
/// ## Description
/// Creates isolines or isobands from an RGB(A) elevation image.
///
/// NOTE: Defaults to the Mapbox elevation data converter [`crate::readers::convert_mapbox_elevation_data`]. However,
/// to use the Terrarium elevation data converter, use [`crate::readers::convert_terrarium_elevation_data`].
///
/// NOTE: Using a buffer with a small padding works, but it only extends the line ends of the isolines.
/// A better method would be to pull in grid data with large padding.
///
/// ## Example
/// ```rust
/// use std::{fs, path::PathBuf};
/// use gistools::{
///     parsers::Buffer,
///     tools::build_contours,
/// };
/// let mut path = PathBuf::from(env!("CARGO_MANIFEST_DIR"));
/// path.push(format!("tests/tools/elevation/fixtures/13_1556_3084.webp"));
/// let elevation_image = fs::read(path).unwrap();
/// let image_buffer = Buffer::from(elevation_image);
///
/// let isolines =
///        build_contours(&image_buffer, None, None, None, None, None);
/// ```
///
/// ## Parameters
/// - `image_data`: image data
/// - `elevation_converter`: the conversion function to convert the pixels to elevation
/// - `step`: the step size for the heightmap. Defaults to 100 meters for the Mapbox elevation data.
/// - `tms_style`: if true, the y position will be inverted
/// - `padding`: The number of pixels that extend around the main data
/// - `tolerance`: The Douglas-Peucker tolerance. Defaults to `1 / 2_096`
///
/// ## Returns
/// The isolines stored in a FeatureCollection
pub fn build_contours(
    image_data: &Buffer,
    elevation_converter: Option<ElevationConverter>,
    step: Option<f64>,
    tms_style: Option<bool>,
    padding: Option<usize>,
    tolerance: Option<f64>,
) -> FeatureCollection {
    let step = step.unwrap_or(100.0);
    let padding = padding.unwrap_or(1);
    let tolerance = tolerance.unwrap_or(1.0 / 2_096.0);
    // create the elevation grid
    let elevation_grid = get_elevation_grid(image_data, elevation_converter, tms_style);
    let IsoGrid { width, height, min, max, elevations } = elevation_grid;
    // setup the thresholds
    let thresholds = isoline_thresholds(min, max, step);

    let mut features: Vec<VectorFeature> = vec![];
    for threshold in thresholds {
        let coordinates =
            build_isobands(&elevations, threshold, width, height, padding, Some(tolerance));
        features.push(VectorFeature::new_wm(
            None,
            (ContourProperties { elev: threshold, elev_ft: threshold * 3.28084 }).into(),
            VectorGeometry::new_multipolygon(coordinates, None),
            None,
        ));
    }

    let mut feature_collection = FeatureCollection::new(None);
    feature_collection.features = features.into_iter().map(Features::VectorFeature).collect();
    feature_collection
}

/// Accumulate, smooth contour rings, assign holes to exterior rings.
/// Based on https://github.com/mbostock/shapefile/blob/v0.6.2/shp/polygon.js
/// @param tolerance - The Douglas-Peucker tolerance. A good default is `1 / 2_096`
pub fn build_isobands(
    values: &[f64],
    threshold: f64,
    width: usize,
    height: usize,
    padding: usize,
    tolerance: Option<f64>,
) -> VectorMultiPolygon {
    let mut polygons: VectorMultiPolygon = vec![];
    let mut holes: VectorMultiLineString = vec![];

    let isorings = build_isorings(values, threshold, width, height, padding, tolerance);
    // Store rings in the correct groups. Skip rings with zero area.
    for ring in isorings {
        let area = ring.area(Some(1.));
        if area == 0. {
            continue;
        } else if area > 0. {
            polygons.push(vec![ring]);
        } else {
            holes.push(ring);
        }
    }
    // sort the holes into their correct polygons
    for hole in holes {
        for poly in &mut polygons {
            if polyline_in_polyline(&hole, &poly[0]) {
                poly.push(hole);
                break;
            }
        }
    }

    polygons
}

/// Marching squares with isolines stitched into rings.
/// Based on https://github.com/topojson/topojson-client/blob/v3.0.0/src/stitch.js
/// @param tolerance - The Douglas-Peucker tolerance. A good default is `1 / 2_096`
pub fn build_isorings(
    values: &[f64],
    threshold: f64,
    width: usize,
    height: usize,
    padding: usize,
    tolerance: Option<f64>,
) -> VectorMultiLineString {
    let tolerance = tolerance.unwrap_or(0.);
    let mut res: VectorMultiLineString = vec![];
    let iwidth = width as isize;
    let mut frag_by_start = BTreeMap::<usize, Rc<RefCell<Fragment>>>::new();
    let mut frag_by_end = BTreeMap::<usize, Rc<RefCell<Fragment>>>::new();
    let mut t0: usize;
    let mut t1: usize;
    let mut t2: usize;
    let mut t3: usize;

    // Special case for the first row (y = -1, t2 = t3 = 0).
    let mut x: isize = -1;
    let mut y: isize = -1;
    t1 = above(values.first().cloned(), threshold);
    MS_LUT[t1 << 1].iter().for_each(|p| {
        stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
    });
    while {
        x += 1;
        x < iwidth - 1
    } {
        t0 = t1;
        t1 = above(values.get((x + 1) as usize).cloned(), threshold);
        MS_LUT[t0 | (t1 << 1)].iter().for_each(|p| {
            stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
        });
    }
    MS_LUT[t1].iter().for_each(|p| {
        stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
    });

    // General case for the intermediate rows.
    while {
        y += 1;
        y < height as isize - 1
    } {
        x = -1;
        t1 = above(values.get((y * iwidth + iwidth) as usize).cloned(), threshold);
        t2 = above(values.get((y * iwidth) as usize).cloned(), threshold);
        MS_LUT[(t1 << 1) | (t2 << 2)].iter().for_each(|p| {
            stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
        });
        while {
            x += 1;
            x < iwidth - 1
        } {
            t0 = t1;
            t1 = above(values.get((y * iwidth + iwidth + x + 1) as usize).cloned(), threshold);
            t3 = t2;
            t2 = above(values.get((y * iwidth + x + 1) as usize).cloned(), threshold);
            MS_LUT[t0 | (t1 << 1) | (t2 << 2) | (t3 << 3)].iter().for_each(|p| {
                stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
            });
        }
        MS_LUT[t1 | (t2 << 3)].iter().for_each(|p| {
            stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
        })
    }

    // Special case for the last row (y = height - 1, t0 = t1 = 0).
    x = -1;
    t2 = (values[(y * iwidth) as usize] >= threshold).into();
    MS_LUT[t2 << 2].iter().for_each(|p| {
        stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
    });
    while {
        x += 1;
        x < iwidth - 1
    } {
        t3 = t2;
        t2 = above(values.get((y * iwidth + x + 1) as usize).cloned(), threshold);
        MS_LUT[(t2 << 2) | (t3 << 3)].iter().for_each(|p| {
            stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
        })
    }
    MS_LUT[t2 << 3].iter().for_each(|p| {
        stitch(p, width, x, y, &mut frag_by_start, &mut frag_by_end, &mut res);
    });

    for ring in &mut res {
        // reverse the ring
        ring.reverse();
        // smooth the edges
        smooth(ring, values, threshold, width as f64, height as f64);
        // convert points to 0->1
        for point in &mut *ring {
            remap(point, width as f64, height as f64, padding as f64);
        }
        // lastly simplify if tolerance is not 0
        if tolerance != 0. {
            // Prep Douglas-Peucker simplification by setting t-values.
            build_sq_dist(ring, 0, ring.len() - 1, tolerance * tolerance);
            // Apply Douglas-Peucker
            *ring = simplify_line(ring, tolerance, false, false);
        }
    }

    res
}

fn stitch(
    line: &[(f64, f64); 2],
    width: usize,
    cx: isize,
    cy: isize,
    start_map: &mut BTreeMap<usize, Rc<RefCell<Fragment>>>,
    end_map: &mut BTreeMap<usize, Rc<RefCell<Fragment>>>,
    results: &mut VectorMultiLineString,
) {
    let start = VectorPoint::from_xy(line[0].0 + cx as f64, line[0].1 + cy as f64);
    let end = VectorPoint::from_xy(line[1].0 + cx as f64, line[1].1 + cy as f64);
    let start_index = index(&start, width as f64);
    let end_index = index(&end, width as f64);

    let f_opt = end_map.get(&start_index).cloned();
    let g_opt = start_map.get(&end_index).cloned();

    if let Some(f) = f_opt {
        if let Some(g) = g_opt {
            end_map.remove(&f.borrow().end);
            start_map.remove(&g.borrow().start);

            if Rc::ptr_eq(&f, &g) {
                // Fragment closes a ring completely
                f.borrow_mut().ring.push(end);
                let finalized_ring = Rc::try_unwrap(f)
                    .unwrap_or_else(|rc| rc.borrow().clone().into())
                    .into_inner()
                    .ring;
                results.push(finalized_ring);
            } else {
                // Merge fragment f and fragment g
                let mut f_borrow = f.borrow_mut();
                let mut g_borrow = g.borrow_mut();

                f_borrow.ring.append(&mut g_borrow.ring);
                f_borrow.end = g_borrow.end;

                let merged_frag =
                    Fragment::new(f_borrow.start, f_borrow.end, f_borrow.ring.clone());
                start_map.insert(f_borrow.start, merged_frag.clone());
                end_map.insert(f_borrow.end, merged_frag);
            }
        } else {
            // Append line segment to existing end map fragment
            end_map.remove(&f.borrow().end);
            f.borrow_mut().ring.push(end);
            f.borrow_mut().end = end_index;
            end_map.insert(end_index, f);
        }
    } else if let Some(f) = start_map.get(&end_index).cloned() {
        if let Some(g) = end_map.get(&start_index).cloned() {
            start_map.remove(&f.borrow().start);
            end_map.remove(&g.borrow().end);

            if Rc::ptr_eq(&f, &g) {
                f.borrow_mut().ring.push(end);
                let finalized_ring = Rc::try_unwrap(f)
                    .unwrap_or_else(|rc| rc.borrow().clone().into())
                    .into_inner()
                    .ring;
                results.push(finalized_ring);
            } else {
                let mut g_borrow = g.borrow_mut();
                let mut f_borrow = f.borrow_mut();

                g_borrow.ring.append(&mut f_borrow.ring);
                g_borrow.end = f_borrow.end;

                let merged_frag =
                    Fragment::new(g_borrow.start, g_borrow.end, g_borrow.ring.clone());
                start_map.insert(g_borrow.start, merged_frag.clone());
                end_map.insert(g_borrow.end, merged_frag);
            }
        } else {
            // Prepend segment to existing start map fragment
            start_map.remove(&f.borrow().start);
            f.borrow_mut().ring.insert(0, start);
            f.borrow_mut().start = start_index;
            start_map.insert(start_index, f);
        }
    } else {
        // Create an entirely new fragment tracking this segment
        let frag = Fragment::new(start_index, end_index, vec![start, end]);
        start_map.insert(start_index, frag.clone());
        end_map.insert(end_index, frag);
    }
}

// Convert a point to an index.
fn index<P: GetXY>(point: &P, width: f64) -> usize {
    (point.x() * 2. + point.y() * (width + 1.) * 4.) as usize
}

// Convert to a number (0 or 1) whether the value is above the threshold.
fn above(x: Option<f64>, value: f64) -> usize {
    x.map(|x| (x >= value) as usize).unwrap_or(0)
}

// Smooth a contour ring.
fn smooth<P: GetXY + SetXY>(
    ring: &mut [P],
    values: &[f64],
    threshold: f64,
    width: f64,
    height: f64,
) {
    ring.iter_mut().for_each(|point| {
        let (x, y) = point.xy();
        let xt = trunc(x);
        let yt = trunc(y);
        let v1 = valid(values.get((yt * width + xt) as usize).cloned());
        if x > 0. && x < width && xt == x {
            point.set_x(smooth1(
                x,
                valid(values.get((yt * width + xt - 1.) as usize).cloned()),
                v1,
                threshold,
            ));
        }
        if y > 0. && y < height && yt == y {
            point.set_y(smooth1(
                y,
                valid(values.get(((yt - 1.) * width + xt) as usize).cloned()),
                v1,
                threshold,
            ));
        }
    });
}

fn valid(v: Option<f64>) -> f64 {
    v.filter(|n| !n.is_nan()).unwrap_or(f64::NEG_INFINITY)
}

fn smooth1(x: f64, v0: f64, v1: f64, value: f64) -> f64 {
    let a = value - v0;
    let b = v1 - v0;
    let d = if a.is_finite() || b.is_finite() { a / b } else { a.signum() / b.signum() };
    if d.is_nan() { x } else { x + d - 0.5 }
}

// Convert the x-y values from width-height scale to 0->1.
// Input data was already offest to represent cell centers ([0.5, 0.5] start)
fn remap<P: GetXY + SetXY>(point: &mut P, width: f64, height: f64, padding: f64) {
    let active_width = width - 2. * padding;
    let active_height = height - 2. * padding;
    point.set_x((point.x() - padding) / active_width);
    point.set_y((point.y() - padding) / active_height);
}
